app/javascript/views/cryptoWalletServices/CryptoWalletServices.jsx

```diff
@@ -0,0 +1,267 @@
+      };
+      newState.solOptions = this.formatOptionData(response, 'SOL')
+      newState.ethOptions = this.formatOptionData(response, 'ETH')
+      
+      axios.get(routes.publishers.cryptoAddressForChannels.index.replace('{channel_id}', this.channel.id)).then((channelResponse) => {
+        newState.currentSolAddress = this.findCurrentAddress('SOL', channelResponse, response)
+        newState.currentEthAddress = this.findCurrentAddress('ETH', channelResponse, response)
+
+        this.setState({ ...newState });
+      });
+    });
+  };
+
+  // crypto connection functions
+
+  getNonce = async () => {
+    return axios.get(routes.publishers.cryptoAddressForChannels.generateNonce.replace('{channel_id}', this.channel.id)).then((response) => {
+      return response.data.nonce;
+    })
+  }
+
+  connectSolanaAddress = async () => {
+    if (!window.solana) {
+      return false;
+    }
+
+    const results = await window.solana.connect();
+    if (results.publicKey) {
+      const pub_key = results.publicKey
+
+      const possibleMatch = this.state.solOptions.filter(sol => sol.value.address === pub_key);
+      if (possibleMatch.length > 0) {
+        this.updateAddress(possibleMatch[0]);
+        return;
+      }
+
+      const nonce = await this.getNonce();
+      if (!nonce) {
+        this.setErrorText('An error occurred. Please try again later.')
+        return;
+      }
+      const encodedMessage = new TextEncoder().encode(nonce)
+      let signedMessage = null
+
+      try {
+        signedMessage = await window.solana.signMessage(encodedMessage, "utf8")
+      } catch (err) {
+        this.setErrorText('Solana connection was interrupted')
+        return;
+      }
+
+      axios({
+        method: 'post',
+        url: routes.publishers.cryptoAddressForChannels.create.replace('{channel_id}', this.channel.id),
+        data: {
+          chain: 'SOL',
+          account_address: pub_key,
+          message: nonce,
+          transaction_signature: bs58.encode(signedMessage.signature),
+        }
+      }).then((response) => {
+        this.handleConnectionResponse(response)
+      });
+    }
+  }
+
+  connectEthereumAddress = async () => {
+    if (window.ethereum) {
+      const web3 = new Web3(window.ethereum);
+      window.ethereum.enable().catch(error => {
+        this.setErrorText('Connect an Ethereum wallet to add an address');
+        return;
+      });
+
+      const addresses = await web3.eth.getAccounts();
+      const address = addresses[0]
+
+      const possibleMatch = this.state.ethOptions.filter(eth => eth.value.address && eth.value.address.toLowerCase() === address.toLowerCase());
+      if (possibleMatch.length > 0) {
+        this.updateAddress(possibleMatch[0]);
+        return;
+      }
+
+      const nonce = await this.getNonce();
+      if (!nonce) {
+        this.setErrorText('An error occurred. Please try again later.')
+        return;
+      }
+
+      web3.eth.sign(nonce, address)
```

Comment:
nit: This library tends to be quite large so if you noticed a slowdown in responsiveness after this PR is merged I can help out with a follow up PR to remove it